[quant][graphmode][fx] Custom module support

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44835 [quant][eagermode] Custom module support
* **#44766 [quant][graphmode][fx] Custom module support**

Summary:
There might be modules that are not symbolically traceable, e.g. LSTM (since it has
input dependent control flows), to support quantization in these cases, user will provide
the corresponding observed and quantized version of the custom module, the observed
custom module with observers already inserted in the module and the quantized version will
have the corresponding ops quantized. And use
```
from torch.quantization import register_observed_custom_module_mapping
from torch.quantization import register_quantized_custom_module_mapping
register_observed_custom_module_mapping(CustomModule, ObservedCustomModule)
register_quantized_custom_module_mapping(CustomModule, QuantizedCustomModule)
```
to register the custom module mappings, we'll also need to define a custom tracer
in order to prevent the custom module from being traced:
```python
class CustomTracer(Tracer):
      def is_leaf_module(self, m):
          return (m.__module__.startswith('torch.nn') and
                    not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, CustomModule)
m = CustomTracer().trace(original_m)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23723455](https://our.internmc.facebook.com/intern/diff/D23723455)

jerryzh168 committed:
[quant][graphmode][fx] Custom module support

Summary:
There might be modules that are not symbolically traceable, e.g. LSTM (since it has
input dependent control flows), to support quantization in these cases, user will provide
the corresponding observed and quantized version of the custom module, the observed
custom module with observers already inserted in the module and the quantized version will
have the corresponding ops quantized. And use
```
from torch.quantization import register_observed_custom_module_mapping
from torch.quantization import register_quantized_custom_module_mapping
register_observed_custom_module_mapping(CustomModule, ObservedCustomModule)
register_quantized_custom_module_mapping(CustomModule, QuantizedCustomModule)
```
to register the custom module mappings, we'll also need to define a custom delegate class
for symbolic trace in order to prevent the custom module from being traced:
```python
class CustomDelegate(DefaultDelegate):
      def is_leaf_module(self, m):
          return (m.__module__.startswith('torch.nn') and
                    not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, CustomModule)
m = symbolic_trace(original_m, delegate_class=CustomDelegate)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Custom module support"


Summary:
There might be modules that are not symbolically traceable, e.g. LSTM (since it has
input dependent control flows), to support quantization in these cases, user will provide
the corresponding observed and quantized version of the custom module, the observed
custom module with observers already inserted in the module and the quantized version will
have the corresponding ops quantized. And use
```
from torch.quantization import register_observed_custom_module_mapping
from torch.quantization import register_quantized_custom_module_mapping
register_observed_custom_module_mapping(CustomModule, ObservedCustomModule)
register_quantized_custom_module_mapping(CustomModule, QuantizedCustomModule)
```
to register the custom module mappings, we'll also need to define a custom tracer
in order to prevent the custom module from being traced:
```python
class CustomTracer(Tracer):
      def is_leaf_module(self, m):
          return (m.__module__.startswith('torch.nn') and
                    not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, CustomModule)
m = CustomTracer().trace(original_m)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23723455](https://our.internmc.facebook.com/intern/diff/D23723455)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Custom module support"


Summary:
There might be modules that are not symbolically traceable, e.g. LSTM (since it has
input dependent control flows), to support quantization in these cases, user will provide
the corresponding observed and quantized version of the custom module, the observed
custom module with observers already inserted in the module and the quantized version will
have the corresponding ops quantized. And use
```
from torch.quantization import register_observed_custom_module_mapping
from torch.quantization import register_quantized_custom_module_mapping
register_observed_custom_module_mapping(CustomModule, ObservedCustomModule)
register_quantized_custom_module_mapping(CustomModule, QuantizedCustomModule)
```
to register the custom module mappings, we'll also need to define a custom tracer
in order to prevent the custom module from being traced:
```python
class CustomTracer(Tracer):
      def is_leaf_module(self, m):
          return (m.__module__.startswith('torch.nn') and
                    not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, CustomModule)
m = CustomTracer().trace(original_m)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23723455](https://our.internmc.facebook.com/intern/diff/D23723455)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Custom module support"


Summary:
There might be modules that are not symbolically traceable, e.g. LSTM (since it has
input dependent control flows), to support quantization in these cases, user will provide
the corresponding observed and quantized version of the custom module, the observed
custom module with observers already inserted in the module and the quantized version will
have the corresponding ops quantized. And use
```
from torch.quantization import register_observed_custom_module_mapping
from torch.quantization import register_quantized_custom_module_mapping
register_observed_custom_module_mapping(CustomModule, ObservedCustomModule)
register_quantized_custom_module_mapping(CustomModule, QuantizedCustomModule)
```
to register the custom module mappings, we'll also need to define a custom tracer
in order to prevent the custom module from being traced:
```python
class CustomTracer(Tracer):
      def is_leaf_module(self, m):
          return (m.__module__.startswith('torch.nn') and
                    not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, CustomModule)
m = CustomTracer().trace(original_m)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23723455](https://our.internmc.facebook.com/intern/diff/D23723455)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Custom module support"


Summary:
There might be modules that are not symbolically traceable, e.g. LSTM (since it has
input dependent control flows), to support quantization in these cases, user will provide
the corresponding observed and quantized version of the custom module, the observed
custom module with observers already inserted in the module and the quantized version will
have the corresponding ops quantized. And use
```
from torch.quantization import register_observed_custom_module_mapping
from torch.quantization import register_quantized_custom_module_mapping
register_observed_custom_module_mapping(CustomModule, ObservedCustomModule)
register_quantized_custom_module_mapping(CustomModule, QuantizedCustomModule)
```
to register the custom module mappings, we'll also need to define a custom tracer
in order to prevent the custom module from being traced:
```python
class CustomTracer(Tracer):
      def is_leaf_module(self, m):
          return (m.__module__.startswith('torch.nn') and
                    not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, CustomModule)
m = CustomTracer().trace(original_m)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23723455](https://our.internmc.facebook.com/intern/diff/D23723455)

[ghstack-poisoned]

supriyar commented on torch/quantization/custom_module_class_mappings.py:
```diff
@@ -0,0 +1,55 @@
+    quantization aware training
+    """
+    assert hasattr(observed_custom_module_class, 'from_float'), 'from_float must be' + \
+        ' defined in observed custom module class'
```
Would be good to provide some details on what is expected in the `from_float` function as well for the Observerd class. Maybe in the docstrings.

jerryzh168 replied:
sure

supriyar commented on test/quantization/test_quantize_fx.py:
```diff
@@ -406,6 +406,122 @@ def forward(self, x):
+                return quantized
+
+        from torch.quantization import register_observed_custom_module_mapping
+        from torch.quantization import register_quantized_custom_module_mapping
```
should we move these imports to the top of file?

jerryzh168 replied:
sure

supriyar commented on torch/quantization/custom_module_class_mappings.py:
```diff
@@ -0,0 +1,55 @@
+OBSERVED_CUSTOM_MODULE_CLASS_MAPPINGS = dict()
```
Are these mappings global?

jerryzh168 replied:
yes, and we have global getter and setter functions listed below

supriyar commented on test/quantization/test_quantize_fx.py:
```diff
@@ -406,6 +406,122 @@ def forward(self, x):
+            def from_observed(cls, observed_module):
+                assert hasattr(observed_module, 'qconfig')
+                assert hasattr(observed_module, 'activation_post_process')
+                observed_module.conv.activation_post_process = \
```
Why does this have to be set by the user? Can't it be done using the qconfig?

jerryzh168 replied:
observer_module.activation_post_process is initialized from qconfig

jerryzh168 replied:
this should be considered using observer_module.activation_post_process to quantize the module, similar as we use mod.activation_post_process in other quantized modules like: https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/modules/conv.py#L256

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -406,6 +406,122 @@ def forward(self, x):
+
+        # define a custom tracer to not trace through the custom module
+
+        class CustomTracer(Tracer):
```
For custom modules, can we get it to work without needing a custom tracer? . We shouldnt ask users to change the  way they symbolically trace their model.

jerryzh168 replied:
i don't think so. fx graph mode is taking a symbolically traced module as input rather than nn.Module as input, so this is outside of our api.

raghuramank100 replied:
In that case, can we atleast only add custom module. Here it looks like the user is expected to know that torch.nn and sequential are existing leaf modules. Can we request the fx team to have an API to register a custom module?

jerryzh168 replied:
that sounds good, I think we can ask fx team to add a custom module api

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -406,6 +406,122 @@ def forward(self, x):
+        m = prepare_static_fx(m, qconfig_dict)
+        # calibration
+        m(data)
+        # all activation observers are inserted in the top level module
```
What is the expected behavior of prepare for custom modules? I am assuming it is a module swap, ie replace with ObservedCustomModule. However, it looks like we are assuming that activation observers are inserted by prepare. Wouldnt it be better to treat custom modules as black boxes? i.e the user specifies where the observers for weights and activations are inserted, the framework just replaces modules.

jerryzh168 replied:
we only insert the activation_post_process for the output, we are treating the module as a blackbox

raghuramank100 replied:
Why not let the user define everything?  If the module has more than one output like LSTMs, what is the expected behavior?

jerryzh168 replied:
we can extend the current code to support multiple outputs and allow user to specify whether each output is quantized or not.
the reason is that we need to know whether these outputs are quantized or not in order to correctly quantize the code following custom module.

currently we assume custom module has one input and one output and both of them are quantized.

raghuramank100 replied:
Can you add a test for LSTMs and ensure that we can handle multiple inputs and outputs, it will likely be the first case for custom modules.

jerryzh168 replied:
yeah i think so, I can add this in a separate PR

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -406,6 +406,122 @@ def forward(self, x):
+                pass
+
+            @classmethod
+            def from_float(cls, float_module):
```
Where are the observers inserted for weights and activations?

jerryzh168 replied:
here we reusing the eager mode conv module, so there is no need to insert weight observers. the observer for activation is inserted outside of the custom module.

raghuramank100 replied:
How would this flow work for QAT? In that case weight observers (fake quant) needs to be inserted at prepare.

jerryzh168 replied:
it will use the qconfig from float_module to instantiate a weight observer and call it in the forward.

jerryzh168 replied:
the flow will be exactly the same

raghuramank100 replied:
Is it correct to say that in general it is the responsibility of the customModule owner to write a Observedclass that contains all the weight +activations observers except for the outputs?

jerryzh168 replied:
yes, that is correct

jerryzh168 committed:
Update on "[quant][graphmode][fx] Custom module support"


Summary:
There might be modules that are not symbolically traceable, e.g. LSTM (since it has
input dependent control flows), to support quantization in these cases, user will provide
the corresponding observed and quantized version of the custom module, the observed
custom module with observers already inserted in the module and the quantized version will
have the corresponding ops quantized. And use
```
from torch.quantization import register_observed_custom_module_mapping
from torch.quantization import register_quantized_custom_module_mapping
register_observed_custom_module_mapping(CustomModule, ObservedCustomModule)
register_quantized_custom_module_mapping(CustomModule, QuantizedCustomModule)
```
to register the custom module mappings, we'll also need to define a custom tracer
in order to prevent the custom module from being traced:
```python
class CustomTracer(Tracer):
      def is_leaf_module(self, m):
          return (m.__module__.startswith('torch.nn') and
                    not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, CustomModule)
m = CustomTracer().trace(original_m)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23723455](https://our.internmc.facebook.com/intern/diff/D23723455)

[ghstack-poisoned]

vkuzo commented on torch/quantization/custom_module_class_mappings.py:
```diff
@@ -0,0 +1,75 @@
+OBSERVED_CUSTOM_MODULE_CLASS_MAPPINGS = dict()
+
+def register_observed_custom_module_mapping(custom_module_class, observed_custom_module_class):
```
optional: how about `registrer_float_mod_to_float_observed_mod_mapping`?

vkuzo commented on torch/quantization/custom_module_class_mappings.py:
```diff
@@ -0,0 +1,75 @@
+
+QUANTIZED_CUSTOM_MODULE_CLASS_MAPPINGS = dict()
+
+def register_quantized_custom_module_mapping(custom_module_class, quantized_custom_module_class):
```
optional: how about `registrer_float_observed_mod_to_quant_mod_mapping`?

vkuzo commented on torch/quantization/custom_module_class_mappings.py:
```diff
@@ -0,0 +1,75 @@
+    QUANTIZED_CUSTOM_MODULE_CLASS_MAPPINGS[custom_module_class] = \
+        quantized_custom_module_class
+
+def get_quantized_custom_module_class(custom_module_class):
```
seems like `custom` could mean either `custom float observed module` or `custom quantized module` in this context, and it might be confusing to the user which one is used where.  Would it be possible to make the names unambiguous?

vkuzo commented on torch/quantization/custom_module_class_mappings.py:
```diff
@@ -0,0 +1,75 @@
+OBSERVED_CUSTOM_MODULE_CLASS_MAPPINGS = dict()
+
+def register_observed_custom_module_mapping(custom_module_class, observed_custom_module_class):
```
maybe we can call `custom_module_class` to be `fp_custom_module_class` or `orig_custom_module_class`, to make it clear that this is a different custom class from observed and quantized?

can we also have a docblock somewhere which clearly explains the (fp) -> (obs) -> (quant) flow in a single paragraph, ideally with a toy example?

jerryzh168 replied:
sure.

for the toy example, I think it's better to do it in a tutorial or the top level quantization doc. I can have a docstring in this file, but not sure how discoverable that is.

vkuzo commented on torch/quantization/custom_module_class_mappings.py:
```diff
@@ -0,0 +1,75 @@
+
+QUANTIZED_CUSTOM_MODULE_CLASS_MAPPINGS = dict()
+
+def register_quantized_custom_module_mapping(custom_module_class, quantized_custom_module_class):
```
should this is `observed_custom_module_class`? otherwise it's unclear if the input should be observed, not observed, or either

jerryzh168 replied:
it's not observed, will change this to float_custom_module_class

vkuzo commented on test/quantization/test_quantize_fx.py:
```diff
@@ -406,6 +408,122 @@ def forward(self, x):
 
+    @skipIfNoFBGEMM
+    def test_custom_module_class(self):
+        class CustomModule(torch.nn.Module):
```
would there be value in adding a more complicated test case which would only work as a custom class?

raghuramank100 replied:
Also, it will be good to add a test case for dynamic quant with custom modules too.

jerryzh168 replied:
can we use LSTM as a test, I'm not sure what to test for.

jerryzh168 replied:
for dynamic quant test, I'll revisit after the mixed mode support change

z-a-f replied:
I think having a complex class in the unittest might be an overkill as it potentially can fail for reasons other than the target of the current unittest. We can instead have multiple unittests for different test scenarios.

raghuramank100 replied:
I think there is value in adding unit tests for the following cases for both static, dynamic and QAT
```python
class CustomModule(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.conv = torch.nn.LSTM(3, 4)

            def forward(self, x, c):
                return self.lstm(x, c)
```
We are not looking for correct implementation of quantized LSTM module here, just making sure that we can handle multiple inputs and outputs for a realistic use case.

Another test can be test a conv + add as a custom module. In this case,  we can check if the custom module support path produces the same output as the default flow.

jerryzh168 replied:
multiple output is not yet supported in this pr, can be done is a separate pr.
sure I can add conv + add test in a separate pr as well

jerryzh168 replied:
I think it would be better to have a real conv + add use case instead of us guessing about what is needed though

vkuzo commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -507,6 +510,25 @@ def convert(self, quantizer, node):
             quantizer.quantized_graph,
             node, quantizer.activation_post_process_map[node.name])
 
+class CustomModule(QuantizeHandler):
```
can we add a docblock?

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -208,6 +213,13 @@ def _prepare(self, model, qconfig_dict, inplace, is_dynamic_quant):
         # match the patterns that will get quantized
         matches = self._find_matches(model.graph, self.modules, self.patterns)
```
perhaps we can move the logic to match custom modules to `_find_matches`, or clearly state in that function that there is also matching logic elsewhere?  Otherwise someone reading the code but not 100% familiar with it can incorrectly assume that `_find_matches` has all the logic to match things.

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -350,6 +376,14 @@ def _convert(self, model, inplace=False, debug=False, is_dynamic_quant=False):
         self.modules = dict(model.named_modules())
 
         matches = self._find_matches(model.graph, self.modules, self.patterns)
```
same here, would it make sense to make all the matching happen in `_find_matches`?

raghuramank100 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -507,6 +510,25 @@ def convert(self, quantizer, node):
+        quantized_custom_module_class = \
+            get_quantized_custom_module_class(observed_custom_module._FLOAT_MODULE)
+        quantized_custom_module = \
+            quantized_custom_module_class.from_observed(observed_custom_module)
```
The code for nn modules only has a from_float that gets called at convert time, this is a better name choice as we call "from_observed", but wondering if we can do anything to reduce inconsistency in names. Any ideas?

jerryzh168 replied:
here we are converting from observed to quantized, are you suggesting to keep the from_float name?

raghuramank100 replied:
No, this is a better name. Not sure what we can do for existing nn modules where we use from_float for both observers, marking as resolved

jerryzh168 replied:
i see, maybe we can refactor current from_float to from_observed as well

jerryzh168 committed:
Update on "[quant][graphmode][fx] Custom module support"


Summary:
There might be modules that are not symbolically traceable, e.g. LSTM (since it has
input dependent control flows), to support quantization in these cases, user will provide
the corresponding observed and quantized version of the custom module, the observed
custom module with observers already inserted in the module and the quantized version will
have the corresponding ops quantized. And use
```
from torch.quantization import register_observed_custom_module_mapping
from torch.quantization import register_quantized_custom_module_mapping
register_observed_custom_module_mapping(CustomModule, ObservedCustomModule)
register_quantized_custom_module_mapping(CustomModule, QuantizedCustomModule)
```
to register the custom module mappings, we'll also need to define a custom tracer
in order to prevent the custom module from being traced:
```python
class CustomTracer(Tracer):
      def is_leaf_module(self, m):
          return (m.__module__.startswith('torch.nn') and
                    not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, CustomModule)
m = CustomTracer().trace(original_m)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23723455](https://our.internmc.facebook.com/intern/diff/D23723455)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Custom module support"


Summary:
There might be modules that are not symbolically traceable, e.g. LSTM (since it has
input dependent control flows), to support quantization in these cases, user will provide
the corresponding observed and quantized version of the custom module, the observed
custom module with observers already inserted in the module and the quantized version will
have the corresponding ops quantized. And use
```
from torch.quantization import register_observed_custom_module_mapping
from torch.quantization import register_quantized_custom_module_mapping
register_observed_custom_module_mapping(CustomModule, ObservedCustomModule)
register_quantized_custom_module_mapping(CustomModule, QuantizedCustomModule)
```
to register the custom module mappings, we'll also need to define a custom tracer
in order to prevent the custom module from being traced:
```python
class CustomTracer(Tracer):
      def is_leaf_module(self, m):
          return (m.__module__.startswith('torch.nn') and
                    not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, CustomModule)
m = CustomTracer().trace(original_m)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23723455](https://our.internmc.facebook.com/intern/diff/D23723455)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Custom module support"


Summary:
There might be modules that are not symbolically traceable, e.g. LSTM (since it has
input dependent control flows), to support quantization in these cases, user will provide
the corresponding observed and quantized version of the custom module, the observed
custom module with observers already inserted in the module and the quantized version will
have the corresponding ops quantized. And use
```
from torch.quantization import register_observed_custom_module_mapping
from torch.quantization import register_quantized_custom_module_mapping
register_observed_custom_module_mapping(CustomModule, ObservedCustomModule)
register_quantized_custom_module_mapping(CustomModule, QuantizedCustomModule)
```
to register the custom module mappings, we'll also need to define a custom tracer
in order to prevent the custom module from being traced:
```python
class CustomTracer(Tracer):
      def is_leaf_module(self, m):
          return (m.__module__.startswith('torch.nn') and
                    not isinstance(m, torch.nn.Sequential)) or \
                    isinstance(m, CustomModule)
m = CustomTracer().trace(original_m)
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23723455](https://our.internmc.facebook.com/intern/diff/D23723455)

[ghstack-poisoned]

z-a-f commented on torch/quantization/fx/quantize.py:
```diff
@@ -18,6 +18,12 @@
 from ..quantization_mappings import (
     get_qat_module_mappings,
 )
+from ..custom_module_class_mappings import (
+    is_custom_module_class,
+    get_observed_custom_module_class,
+    mark_observed_custom_module,
+    is_observed_custom_module,
+)
```
Can we split these into separate lines -- it has no performance impact, and in case of the refactoring it's easier to reloacte

jerryzh168 replied:
you mean the import? it's already in separate lines

z-a-f reviewed (approved):
LGTM
